### PR TITLE
Whisper live transcription

### PR DESCRIPTION
**Pull request by rpurdel**

### CODE REVIEW

File: src/main/java/org/jitsi/jigasi/transcription/WhisperWebsocket.java [hunk 1]
```diff
@@ -0,0 +1,339 @@
...
+import java.util.*;
+import java.util.concurrent.CompletableFuture;
+
+
+@WebSocket
+public class WhisperWebsocket {
+
+    private Session wsSession;
+
+    private HashMap<String, Participant> participants = new HashMap<>();
```
Comment (bgrozev) [inline src/main/java/org/jitsi/jigasi/transcription/WhisperWebsocket.java:30]: I think access to these fields isn't thread safe. Does it need to be?

File: src/main/java/org/jitsi/jigasi/transcription/WhisperWebsocket.java [hunk 2]
```diff
@@ -0,0 +1,339 @@
...
+    }
+
+    public void sendAudio(String participantId, Participant participant, ByteBuffer audio) {
+        addParticipantIfNotExists(participantId, participant);
+        try
+        {
+            logger.debug("Sending audio for " + participantId);
+            wsSession.getRemote().sendBytes(buildPayload(participantId, participant, audio));
+        }
+        catch (NullPointerException e)
```
Comment (bgrozev) [inline src/main/java/org/jitsi/jigasi/transcription/WhisperWebsocket.java:297]: Can we add a null check instead?

**Review by bgrozev — COMMENTED** [review]
I added some comments for things that I noticed, but I didn't read thoroughly (I leave that to @damencho). Notably, I think there are some thread safety issues (unless I'm missing a reason why thread safety is not needed).

File: src/main/java/org/jitsi/jigasi/TranscriptionGatewaySession.java [hunk 3]
```diff
@@ -631,13 +631,26 @@ private void sendMessageToRoom(String messageString)
...
+    }
+
     /**
      * Send a {@link TranscriptionResult} to the {@link ChatRoom}
      *
      * @param result the {@link TranscriptionResult} to send
      */
     private void sendTranscriptionResultToRoom(TranscriptionResult result)
     {
+        refreshRoom();
```
Comment (damencho) [inline src/main/java/org/jitsi/jigasi/TranscriptionGatewaySession.java:653]: Why is this needed?

File: src/main/java/org/jitsi/jigasi/transcription/WhisperTsAudioSilenceCaptureDevice.java [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+class WhisperTsAudioSilenceCaptureDevice extends AbstractPushBufferCaptureDevice {
+    private static final long CLOCK_TICK_INTERVAL = 20L;
+    private static final Format[] SUPPORTED_FORMATS;
+    private final boolean clockOnly;
+
+    private final static Logger logger
+            = Logger.getLogger(WhisperTsAudioSilenceCaptureDevice.class);
+
+    WhisperTsAudioSilenceCaptureDevice(boolean clockOnly) {
+        this.clockOnly = clockOnly;
```
Comment (damencho) [inline src/main/java/org/jitsi/jigasi/transcription/WhisperTsAudioSilenceCaptureDevice.java:25]: Please run the check style, this for sure should fail ({ should be on new line). If it doesn't fail we need to fix the check style config.

File: src/main/java/org/jitsi/jigasi/TranscriptionGatewaySession.java (hunk 3 above)
Comment (rpurdel) [reply src/main/java/org/jitsi/jigasi/TranscriptionGatewaySession.java:653]: Because otherwise I get this when starting a second transcription and no transcription is sent to the chatRoom

```
SEVERE: Cannot send message as chatRoom is null
```

File: src/main/java/org/jitsi/jigasi/transcription/TranscribingAudioMixerMediaDevice.java [hunk 5]
```diff
@@ -18,6 +18,25 @@
 package org.jitsi.jigasi.transcription;
 
 import org.jitsi.impl.neomedia.device.*;
+import org.jitsi.impl.neomedia.format.MediaFormatImpl;
```
[the remark this answers is not in the document]
Comment (damencho) [reply src/main/java/org/jitsi/jigasi/transcription/TranscribingAudioMixerMediaDevice.java]: @rpurdel You should update everything to use the star imports, as the rest of the code does.

File: src/main/java/org/jitsi/jigasi/transcription/WhisperWebsocket.java [hunk 6]
```diff
@@ -0,0 +1,339 @@
...
+        Date now = new Date(nowMillis);
+        KeyFactory kf = KeyFactory.getInstance("RSA");
+        PKCS8EncodedKeySpec keySpecPKCS8 = new PKCS8EncodedKeySpec(Base64.getDecoder().decode(privateKey));
+        PrivateKey finalPrivateKey = kf.generatePrivate(keySpecPKCS8);
+        JwtBuilder builder = Jwts.builder()
+                .setHeaderParam("kid", privateKeyName)
+                .setIssuedAt(now)
+                .setIssuer("jigasi")
+                .signWith(SignatureAlgorithm.RS256, finalPrivateKey);
+        long expires = nowMillis + (60 * 60 * 1000);
```
[the remark this answers is not in the document]
Comment (rpurdel) [reply src/main/java/org/jitsi/jigasi/transcription/WhisperWebsocket.java:114]: no, the JWT is only used for the initial connection

[the remark this answers is not in the document]
Comment (rpurdel) [reply src/main/java/org/jitsi/jigasi/transcription/WhisperWebsocket.java:114]: I will make it expire sooner

File: src/main/java/org/jitsi/jigasi/transcription/WhisperConnectionPool.java [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+    private HashMap<String, WhisperWebsocket> connections = new HashMap<>();
+
+    /**
+     * Gets a connection if it exists, creates one if it doesn't.
+     * @param roomId
+     * @param participantId
+     * @return
+     * @throws Exception
+     */
+    public WhisperWebsocket getConnection(String roomId, String participantId) throws Exception {
```
[the remark this answers is not in the document]
Comment (rpurdel) [reply src/main/java/org/jitsi/jigasi/transcription/WhisperConnectionPool.java:71]: Converted the hashmaps to concurrent hashmaps

File: src/main/java/org/jitsi/jigasi/health/SipHealthPeriodicChecker.java [hunk 8]
```diff
@@ -21,6 +21,7 @@
 import net.java.sip.communicator.service.protocol.event.*;
 import net.java.sip.communicator.service.protocol.media.*;
 import org.apache.commons.lang3.StringUtils;
+import org.jitsi.impl.neomedia.device.AudioSilenceMediaDevice;
```
[the remark this answers is not in the document]
Comment (damencho) [reply src/main/java/org/jitsi/jigasi/health/SipHealthPeriodicChecker.java:24]: @rpurdel can you please fix everywhere to use package imports (star imports)?

**Comment by damencho** [comment]
Have you ran the check style? It should complain for all those `{` that are not on new line.

**Comment by damencho** [comment]
How are you building the project? `mvn install` runs the checkstyle:
```
[INFO] --- maven-checkstyle-plugin:3.1.2:check (default) @ jigasi ---
[INFO] You have 0 Checkstyle violations.
```

File: src/main/java/org/jitsi/jigasi/transcription/WhisperWebsocket.java (hunk 2 above)
Comment (rpurdel) [reply src/main/java/org/jitsi/jigasi/transcription/WhisperWebsocket.java]: Added it

File: src/main/java/org/jitsi/jigasi/transcription/Transcriber.java [hunk 9]
```diff
@@ -29,6 +29,9 @@
 import org.jivesoftware.smack.packet.Presence;
 
 import javax.media.Buffer;
+import javax.media.Format;
+import javax.media.format.AudioFormat;
+import javax.media.protocol.CaptureDevice;
 import javax.media.rtp.*;
```
Comment (damencho) [inline src/main/java/org/jitsi/jigasi/transcription/Transcriber.java:35]: Can you change and these imports, please?

File: src/main/java/org/jitsi/jigasi/transcription/WhisperAudioSilenceCaptureDevice.java [hunk 10]
```diff
@@ -0,0 +1,397 @@
...
+import org.jitsi.impl.neomedia.codec.AbstractCodec2;
+import org.jitsi.impl.neomedia.jmfext.media.protocol.*;
+import org.jitsi.impl.neomedia.jmfext.media.renderer.audio.AbstractAudioRenderer;
+
+import javax.media.*;
+import javax.media.control.FormatControl;
+import javax.media.format.AudioFormat;
+import javax.media.protocol.BufferTransferHandler;
+import java.io.IOException;
+import java.util.Arrays;
```
Comment (damencho) [inline src/main/java/org/jitsi/jigasi/transcription/WhisperAudioSilenceCaptureDevice.java:29]: And these?

File: src/main/java/org/jitsi/jigasi/transcription/WhisperAudioSilenceMediaDevice.java [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jitsi.jigasi.transcription;
+
+import org.jitsi.impl.neomedia.device.*;
+import org.jitsi.service.neomedia.MediaDirection;
```
Comment (damencho) [inline src/main/java/org/jitsi/jigasi/transcription/WhisperAudioSilenceMediaDevice.java:21]: Here?

File: src/main/java/org/jitsi/jigasi/transcription/WhisperConnectionPool.java [hunk 12]
```diff
@@ -0,0 +1,134 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jitsi.jigasi.transcription;
+
+import org.jitsi.utils.logging.Logger;
+
+import java.io.IOException;
```
Comment (damencho) [inline src/main/java/org/jitsi/jigasi/transcription/WhisperConnectionPool.java:22]: And this?

File: src/main/java/org/jitsi/jigasi/transcription/WhisperTranscriptionService.java [hunk 13]
```diff
@@ -0,0 +1,187 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jitsi.jigasi.transcription;
+
+import org.eclipse.jetty.websocket.api.*;
+import org.jitsi.impl.neomedia.device.AudioMixerMediaDevice;
+import org.jitsi.impl.neomedia.device.ReceiveStreamBufferListener;
```
Comment (damencho) [inline src/main/java/org/jitsi/jigasi/transcription/WhisperTranscriptionService.java:22]: More.

File: src/main/java/org/jitsi/jigasi/transcription/WhisperWebsocket.java [hunk 14]
```diff
@@ -0,0 +1,378 @@
...
+
+import java.io.IOException;
+import java.net.URI;
+import java.nio.ByteBuffer;
+import java.security.*;
+import java.security.spec.*;
+import java.time.*;
+import java.util.*;
+import java.util.concurrent.*;
+
```
Comment (damencho) [inline src/main/java/org/jitsi/jigasi/transcription/WhisperWebsocket.java:36]: Here

**Review by damencho — DISMISSED** [review]

File: src/main/java/org/jitsi/jigasi/TranscriptionGateway.java [hunk 15]
```diff
@@ -18,10 +18,13 @@
...
-import org.jitsi.jigasi.transcription.action.*;
-import org.jitsi.utils.logging.*;
+import org.jitsi.jigasi.transcription.action.ActionServicesHandler;
+import org.jitsi.utils.logging.Logger;
 import org.json.*;
-import org.osgi.framework.*;
+import org.osgi.framework.BundleContext;
+
+import java.io.*;
+import java.net.*;
```
Comment (damencho) [inline src/main/java/org/jitsi/jigasi/TranscriptionGateway.java:28]: Can you fix the imports.

File: src/main/java/org/jitsi/jigasi/TranscriptionGateway.java [hunk 16]
```diff
@@ -91,29 +100,55 @@ public void stop()
...
+                    }
+                    inputStream.close();
+                    logger.info("+++++++++ Received json " + responseBody);
+                    JSONObject obj = new JSONObject(responseBody.toString());
+                    transcriberClass = obj.getString("transcriber");
+                    logger.info("+++++++++ Using " + transcriberClass + " as the transcriber class.");
+                }
+                else
+                {
+                    logger.warn("+++++++++ Got HTTP status code " + responseCode);
```
Comment (damencho) [inline src/main/java/org/jitsi/jigasi/TranscriptionGateway.java]: Can you drop the debug logging.

File: src/main/java/org/jitsi/jigasi/TranscriptionGateway.java [hunk 17]
```diff
@@ -91,29 +100,55 @@ public void stop()
...
+                }
+                conn.disconnect();
+                return transcriberClass;
             }
-            catch (JSONException ex)
+            catch (Exception ex)
             {
-                logger.warn("Could not find '" + tenant + "' tenant in custom transcription service JSON property.");
-                return null;
+                logger.error("+++++++++++++ Could not retrieve transcriber from remote URL." + ex);
```
Comment (damencho) [inline src/main/java/org/jitsi/jigasi/TranscriptionGateway.java]: Can you drop the debug logging.

File: src/main/java/org/jitsi/jigasi/TranscriptionGateway.java [hunk 18]
```diff
@@ -91,29 +100,55 @@ public void stop()
...
+                    BufferedReader inputStream = new BufferedReader(
+                            new InputStreamReader(conn.getInputStream()));
+                    String inputLine;
+                    StringBuilder responseBody = new StringBuilder();
+                    while ((inputLine = inputStream.readLine()) != null)
+                    {
+                        responseBody.append(inputLine);
+                    }
+                    inputStream.close();
+                    logger.info("+++++++++ Received json " + responseBody);
```
Comment (damencho) [inline src/main/java/org/jitsi/jigasi/TranscriptionGateway.java]: debug logs

File: src/main/java/org/jitsi/jigasi/TranscriptionGateway.java [hunk 19]
```diff
@@ -91,29 +100,55 @@ public void stop()
...
+                    JSONObject obj = new JSONObject(responseBody.toString());
+                    transcriberClass = obj.getString("transcriber");
+                    logger.info("+++++++++ Using " + transcriberClass + " as the transcriber class.");
```
Comment (damencho) [inline src/main/java/org/jitsi/jigasi/TranscriptionGateway.java]: debug logs

**Review by damencho — CHANGES_REQUESTED** [review]

**Review by damencho — APPROVED** [review]
